flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/join/stream/StreamingJoinOperator.java

```diff
@@ -0,0 +1,465 @@
+						}
+					}
+					// send +[record+other]s
+					outRow.setHeader(BaseRowUtil.ACCUMULATE_MSG);
+					for (BaseRow other : associatedRecords.getRecords()) {
+						output(input, other, inputIsLeft);
+					}
+				}
+				// skip when there is no matched rows on the other side
+			}
+		} else { // input record is retract
+			// state.retract(record)
+			inputSideStateView.retractRecord(input);
+			if (associatedRecords.isEmpty()) { // there is no matched rows on the other side
+				if (inputIsOuter) { // input side is outer
+					// send -[record+null]
+					outRow.setHeader(BaseRowUtil.RETRACT_MSG);
+					outputNullPadding(input, inputIsLeft);
+				}
+				// nothing to do when input side is not outer
+			} else { // there are matched rows on the other side
+				// send -[record+other]s
+				outRow.setHeader(BaseRowUtil.RETRACT_MSG);
+				for (BaseRow other : associatedRecords.getRecords()) {
+					output(input, other, inputIsLeft);
+				}
+				// if other side is outer
+				if (otherIsOuter) {
+					OuterJoinRecordStateView otherSideOuterStateView = (OuterJoinRecordStateView) otherSideStateView;
+					for (OuterRecord outerRecord : associatedRecords.getOuterRecords()) {
+						if (outerRecord.numOfAssociations == 1) {
+							// send +[null+other]
+							outRow.setHeader(BaseRowUtil.ACCUMULATE_MSG);
+							outputNullPadding(outerRecord.record, !inputIsLeft);
+						} // nothing else to do when number of associations > 1
+						// otherState.update(other, old - 1)
+						otherSideOuterStateView.updateNumOfAssociations(outerRecord.record, outerRecord.numOfAssociations - 1);
+					}
+				}
+			}
+		}
+	}
+
+	// -------------------------------------------------------------------------------------
+
+	private void output(BaseRow inputRow, BaseRow otherRow, boolean inputIsLeft) {
+		if (inputIsLeft) {
+			outRow.replace(inputRow, otherRow);
+		} else {
+			outRow.replace(otherRow, inputRow);
+		}
+		collector.collect(outRow);
+	}
+
+	private void outputNullPadding(BaseRow row, boolean isLeft) {
+		if (isLeft) {
+			outRow.replace(row, rightNullRow);
+		} else {
+			outRow.replace(leftNullRow, row);
+		}
+		collector.collect(outRow);
+	}
+
+	// ----------------------------------------------------------------------------------------
+	// Utility Classes
+	// ----------------------------------------------------------------------------------------
+
+	private class JoinConditionWithNullFilters implements JoinCondition {
+
+		final JoinCondition backingJoinCondition;
+
+		private JoinConditionWithNullFilters(JoinCondition backingJoinCondition) {
+			this.backingJoinCondition = backingJoinCondition;
+		}
+
+		@Override
+		public boolean apply(BaseRow left, BaseRow right) {
+			if (nullFilterKeys.length != 0) { // is not null safe, return false if any null exists
+				BaseRow joinKey = (BaseRow) getCurrentKey();
+				for (int idx : nullFilterKeys) {
```

Comment:
Is joinKey is alway `BinaryRow`?
If it is, I recommend doing something like:
`filterAllNulls ? buildKey.anyNull() : buildKey.anyNull(nullFilterKeys)`
in `BinaryHashTable.applyCondition`.
Because `BinaryRow.anyNull` is more efficient.

Reply:
Thanks for the tip.